content/posts/0020-manually-testing-changes-to-paketo-buildpacks.md

```diff
@@ -0,0 +1,249 @@
+---
+title: Manually testing changes to a Paketo Buildpack
+date: "2023-05-23"
+slug: manually-testing-changes-to-paketo-buildpacks
+author: anthonydahanne
+---
+
+*While this article is mainly about how to manually test Java family Paketo buildpacks, still, most of the instructions should work fine with other Paketo buildpacks; and even non Paketo buildpacks!*
+
+Suppose that you started working on a new feature for the [`spring-boot`](https://github.com/paketo-buildpacks/spring-boot/) buildpack...
+
+## Unit tests first 
+
+Of course, during your development, you will make sure that you did not break any tests, regularly running `go test -v ./...` but you would also add new tests covering your new code; the classic combination `go test -v ./... -coverprofile=coverage.out` followed with `go tool cover -html=coverage.out` will allow you to verify you did not forget to test some errors or edge cases.
+
+## Build and try the buildpack out before publishing it
+
+If the buildpack you are working on provides some integration tests (or end to end tests), I suggest you try and add some use cases to those tests.
+
+If not, then you probably want to try it out.
+
+### Build the Paketo Buildpack
+Most Paketo buildpacks provide a `scripts/build.sh` build script that you can use to build the `detect` and `build` binaries (luckily the required `buildpack.toml` is already at the root of the project).
+
+Just run the script and then check you have your buildpacks `bin` and `detect` executable files generated:
+
+```shell
+> scripts/build.sh
```

Comment:
I'm not sure you technically need to, but I would suggest running `create-package` instead of running the script directly. That will take care of setting a version.

There are also some [scripts floating around](https://gist.github.com/dmikusa/8e45bb550f194a0606e516372a2f26b9#file-config-fish-L163) (I think @pivotal-david-osullivan might have a bash version) that wrap all this up and package up a dev buildpack image.

I'm conflicted about recommending one of these scripts because they're not official, but I do feel like they make the process easier. Idk, maybe we could create an official script for this?

Reply:
you're right about the convenience of using `create-package` , but then again: another tool that needs to be installed when just bash and go tool chain and pack cli are enough and probably installed on any buildpacks users machine.

I don't want to explain in this article the `go install -ldflags="-s -w" github.com/paketo-buildpacks/libpak/cmd/create-package@latest` command which additionally makes those instruction even less suitable for other families buildpacks.

Reply:
That's fair. It might be less awkward to edit buildpack.toml and put in a version number. Otherwise you get the literal string we use as a placeholder in the output. I don't think that will cause issues, but it looks odd so minimally it would be worth calling out why you see that if you don't suggest editing it.